api/src/main/java/org/apache/iceberg/expressions/BoundAggregate.java

```diff
@@ -44,4 +57,85 @@ public Type type() {
       return term().type();
     }
   }
+
+  public String columnName() {
+    if (op() == Operation.COUNT_STAR) {
+      return "*";
+    } else {
+      return ref().name();
+    }
+  }
+
+  public String describe() {
+    switch (op()) {
+      case COUNT_STAR:
+        return "count(*)";
+      case COUNT:
+        return "count(" + ExpressionUtil.describe(term()) + ")";
+      case MAX:
+        return "max(" + ExpressionUtil.describe(term()) + ")";
+      case MIN:
+        return "min(" + ExpressionUtil.describe(term()) + ")";
+      default:
+        throw new UnsupportedOperationException("Unsupported aggregate type: " + op());
+    }
+  }
+
+  <V> V safeGet(Map<Integer, V> map, int key) {
+    return safeGet(map, key, null);
+  }
+
+  <V> V safeGet(Map<Integer, V> map, int key, V defaultValue) {
+    if (map != null) {
+      return map.getOrDefault(key, defaultValue);
+    }
+
+    return null;
+  }
+
+  interface Aggregator<R> {
+    void update(StructLike struct);
+
+    void update(DataFile file);
+
+    R result();
+  }
+
+  abstract static class NullSafeAggregator<T, R> implements Aggregator<R> {
+    private final BoundAggregate<T, R> aggregate;
+    private boolean isNull = false;
+
+    NullSafeAggregator(BoundAggregate<T, R> aggregate) {
+      this.aggregate = aggregate;
+    }
+
+    protected abstract void update(R value);
+
+    protected abstract R current();
+
+    @Override
+    public void update(StructLike struct) {
+      if (!isNull) {
+        R value = aggregate.eval(struct);
+        update(value);
+      }
+    }
+
+    @Override
+    public void update(DataFile file) {
+      if (!isNull) {
+        R value = aggregate.eval(file);
+        update(value);
```

Comment:
I removed the [null check](https://github.com/apache/iceberg/blob/97d657cc4485da9f10b74d48cc3c4c52cadffcef/api/src/main/java/org/apache/iceberg/expressions/BoundAggregate.java#L124) and did it in `update`. e.g. [MaxAggregate.update]( https://github.com/apache/iceberg/blob/349daf83882655d13d591807fe7b9ded268ec5d4/api/src/main/java/org/apache/iceberg/expressions/MaxAggregate.java#L62)

Reply:
Was this needed for correctness? I'm not sure I understand why you'd need to move it. The `NullSafeAggregator` was intended to avoid needing to handle null in the `update` methods. It also used to keep `isNull` set correctly so that any aggregate that is null would stop calling `eval` and updating.

Reply:
If one data file  evaluates `null`, I think we still want to evaluate the rest of the data files. For example,
```
CREATE TABLE test (id LONG, data INT) USING iceberg PARTITIONED BY (id);

INSERT INTO TABLE test VALUES (1, null), (1, null), (2, 33), (2, 44), (3, 55), (3, 66);

SELECT max(data) FROM test;
```
For `max(data)`, the first data file evaluates null, I think we still want to evaluate the rest of the data files to get the max value `66` for `max(data)`.

Reply:
I see. In that case, I think we need to change `isNull` to `hasValue` and return a boolean from `update(R)`.

The intent here was to signal when there is not enough information to produce a value. When there isn't, then the result value should be `null`, and we can skip pulling values out of rows or data files because we don't have enough information.

For example, if we are processing 3 Parquet files and 1 Avro file, the Avro file may not have a max value. Rather than giving a partial max from the 3 Parquet files, we need `update(avroFile)` to return `hasValue = false` so that we stop aggregating.

You're right that this needs to change from my original version, which assumed any null value signaled that there was no maximum. If we know that a file contains only null values, then we can skip it even if it doesn't have an upper bound. Similarly, if we get a null value from a row then we can skip it.

Reply:
I changed `isNull` to `hasValue`. I have also added a flag `canPushDown` in `BoundAggregate` to indicate if this aggregate can be pushed down. I think I need a way to differentiate the `null`: is the `null` due to stats not available (e.g. complex type) or due to the value is null, so I added this flag.